rllib/utils/exploration/noveld.py

```diff
@@ -0,0 +1,463 @@
+    used. Here the distillation error is used.
+    """
+
+    def __init__(
+        self,
+        action_space: Space,
+        *,
+        framework: str,
+        model: ModelV2,
+        embed_dim: int = 128,
+        distill_net_config: Optional[ModelConfigDict] = None,
+        lr: float = 1e-3,
+        alpha: float = 0.5,
+        beta: float = 0.0,
+        intrinsic_reward_coeff: float = 5e-3,
+        normalize: bool = True,
+        random_timesteps: int = 10000,
+        sub_exploration: Optional[FromConfigSpec] = None,
+        **kwargs,
+    ):
+        """Initializes a NovelD exploration scheme.
+
+        Args:
+            action_space: The action space of the environment.
+            framework: The framework used to train the model.
+            model: The model used to train the model.
+        """
+        if not isinstance(action_space, (Discrete, MultiDiscrete)):
+            raise ValueError(
+                "Only (Multi)Discrete action spaces supported for NovelD so far!"
+            )
+
+        super().__init__(action_space, framework=framework, model=model, **kwargs)
+
+        if self.policy_config["num_workers"] != 0:
+            raise ValueError(
+                "NovelD exploration currently does not support parallelism."
+                " `num_workers` must be 0!"
+            )
+
+        # Try to import xxhash.
+        try:
+            import xxhash
+
+            self._hash_state = self._xxhash_state
+            logger.info(
+                "Initializing NovelD: Found `xxhash`. Using it for hashing "
+                "visited states in exploration as it offers higher performance "
+                "and can provide hashing for larger observation spaces."
+            )
+        except ImportError as e:
+            self._hash_state = self._defaulthash_state
+            logger.warning(
+                "Initializing NovelD: `xxhash` not found. Falling back to "
+                "default hashing. If you want to install `xxhash` use "
+                "`pip install xxhash`. `xxhash` shows higher performance and "
+                "can provide hashing for larger observation spaces."
+            )
+
+        self.embed_dim = embed_dim
+        # In case no configuration is passed in, use the Policy's model config.
+        # it has the same input dimensions as needed for the distill network.
+        if distill_net_config is None:
+            distill_net_config = self.policy_config["model"].copy()
+        self.distill_net_config = distill_net_config
+        self.lr = lr
+        self.alpha = alpha
+        self.beta = beta
+        self.intrinsic_reward_coeff = intrinsic_reward_coeff
+        self.normalize = normalize
+
+        self._state_counts = {}
+        self._state_counts_total = 0.0
+        self._state_counts_avg = 0.0
+        if self.normalize:
+            from ray.rllib.utils.exploration.random_encoder import MovingMeanStd
+
+            self._moving_mean_std = MovingMeanStd()
```

Comment:
Can we set this to None, if self.normalize=False?

Reply:
I have set the default to `self._moving_mean_std=None`

Reply:
Done